src/v/raft/id_allocator_stm.h

```diff
@@ -0,0 +1,174 @@
+public:
+    struct stm_allocation_result {
+        int64_t id;
+        raft::errc raft_status{raft::errc::success};
+    };
+
+    explicit id_allocator_stm(
+      ss::logger&, raft::consensus*, config::configuration&);
+
+    ss::future<> start() final;
+
+    ss::future<stm_allocation_result>
+    allocate_id_and_wait(model::timeout_clock::time_point timeout);
+
+private:
+    struct sequence_id {
+        model::run_id run_id;
+        model::node_id node_id;
+        int64_t tick;
+
+        bool operator==(const sequence_id& other) const {
+            return run_id == other.run_id && node_id == other.node_id
+                   && tick == other.tick;
+        }
+
+        bool operator!=(const sequence_id& other) const {
+            return !(*this == other);
+        }
+
+        template<typename H>
+        friend H AbslHashValue(H h, const sequence_id& seq) {
+            return H::combine(std::move(h), seq.run_id, seq.node_id, seq.tick);
+        }
+    };
+
+    struct log_allocation_result {
+        sequence_id seq;
+        int64_t base;
+        int64_t range;
+        raft::errc raft_status{raft::errc::success};
+        model::offset offset;
+    };
+
+    struct allocation_cmd {
+        static constexpr uint8_t record_key = 0;
+        sequence_id seq;
+        int64_t range;
+    };
+
+    struct prepare_truncation_cmd {
+        static constexpr uint8_t record_key = 1;
+        sequence_id seq;
+    };
+
+    struct execute_truncation_cmd {
+        static constexpr uint8_t record_key = 2;
+        model::offset prepare_offset;
+        int64_t state;
+    };
+
+    struct cached_allocation_cmd {
+        model::offset offset;
+        allocation_cmd cmd;
+    };
+
+    ss::future<> process(model::record_batch&& b);
+    void execute(model::offset offset, allocation_cmd c);
+
+    ss::future<> apply(model::record_batch b) override;
+    ss::future<result<raft::replicate_result>> replicate(model::record_batch&&);
+    ss::future<log_allocation_result> replicate_and_wait(
+      allocation_cmd, model::timeout_clock::time_point, sequence_id);
+    ss::future<bool> replicate_and_wait(
+      prepare_truncation_cmd, model::timeout_clock::time_point, sequence_id);
+
+    config::configuration& _config;
+    int64_t _last_seq_tick;
+    std::optional<model::run_id> _run_id;
+
+    raft::consensus* _c;
+    absl::flat_hash_map<sequence_id, expiring_promise<log_allocation_result>>
+      _promises;
+    absl::flat_hash_map<sequence_id, expiring_promise<bool>> _prepare_promises;
+
+    int64_t _state{0};
+    int64_t _processed{0};
+    int64_t _last_allocated_base{0};
+    int64_t _last_allocated_range{0};
+    bool _should_cache{false};
+    model::offset _prepare_offset{0};
+    int64_t _prepare_state{0};
+    std::vector<cached_allocation_cmd> _cache;
+
+    static inline constexpr model::record_batch_type id_allocator_stm_batch_type
+      = model::record_batch_type(8);
```

Comment:
`model::record_batch_type(8);` this looks like it might be duplicated? We should use the named types from a central location. I think I saw this done in the previous commit?